### PR TITLE
Queue | Attorney check out | Enter issue dispositions

### PR DESCRIPTION
**Pull request by dannysellers**
Connects #4471, https://github.com/department-of-veterans-affairs/appeals-design-research/issues/684

### Acceptance Criteria
- [x] Verify that there are the following table headers: "Issues", "Actions", "Dispositions" in bold
- [x] Verify that the issue display under the issues header in the following hierarchy: 
   - Program
   - Issue
   - Levels 1-3 (i.e. 5237 - Lumbosacral or cervical strain
   - Notes (if there)
- [x] Verify that underneath the "Action" label there is a link that says "Edit issue" that ~opens the "Edit issue" page~ (#4472)
- [x] Verify that underneath the "Dispositions" label there is a dropdown labeled "Select Dispositions" 
   - Attorney can select a disposition for each issue:
      - "1" => "Allowed",
      - "3" => "Remanded",
      - "4" => "Denied",
      - "5" => "Vacated",
      - "6" => "Dismissed, Other",
      - "8" => "Dismissed, Death",
      - "9" => "Withdrawn",
   - Issue dispositions update in VACOLS
- [x] If attorney selects 'Vacated', another check box appears with copy "Vacate and readjudicate this issue. Automatically duplicate vacated issue."
- [x] If attorney selects that checkbox, verify that the issue is duplicated in VACOLS
    - From Jed: "If an attorney selects disposition 5-Vacate the prompt above displays, if they answer ‘Yes’ it will duplicate the vacated issue with a null disposition."
- [x] Verify that at the end of the last issue there is a link called "Add issue" that ~opens the "Add issue" page~ (#4477)
- [x] Verify that at the bottom left of the app canvas there is a link that navigates the user back to the work product page with the following copy "Cancel and go back to Ready for Judge Review - Select Work Product" 
- [x] Verify that there is a button to the bottom right of the App Canvas that says "Finish dispositions"
- [x] Verify that the user cannot advance to the next screen unless all issues have dispositions filled out.

### CODE REVIEW

File: client/app/queue/reducers.js [hunk 1]
```diff
@@ -69,6 +82,58 @@ const workQueueReducer = (state = initialState, action = {}) => {
...
+    });
+  case ACTIONS.CANCEL_EDITING_OBJECT:
+    return update(state, {
+      pendingChanges: {
+        [action.payload.type]: {
+          $unset: action.payload.vacolsId
+        }
+      }
+    });
+  case ACTIONS.UPDATE_APPEAL_ISSUE: {
```
Comment (dannysellers) [inline client/app/queue/reducers.js]: Block needed here to get around `eslint`'s [`no-case-declarations`](https://eslint.org/docs/rules/no-case-declarations)

File: client/app/queue/reducers.js [hunk 2]
```diff
@@ -59,14 +74,72 @@ const workQueueReducer = (state = initialState, action = {}) => {
...
+    const issueIdx = _.findIndex(issues, (issue) => issue.id === action.payload.issueId);
+
+    return update(state, {
+      pendingChanges: {
+        appeals: {
+          [action.payload.appealId]: {
+            attributes: {
+              issues: {
+                [issueIdx]: {
+                  $merge: action.payload.attributes
```
Comment (mdbenjam) [inline client/app/queue/reducers.js]: It's generally a bad practice to have very nested state. However, are you trying to keep this in a particular structure to make posting it back to the server easier?

File: spec/feature/queue/queue_spec.rb [hunk 3]
```diff
@@ -277,5 +277,70 @@
...
+        table_rows[0..0].each do |row|
+          row.find(".Select-control").click
+          row.find("div[id$='--option-1']").click
+        end
+
+        safe_click("#finish-dispositions")
+
+        table_rows[1..-1].each do |row|
+          dropdown_border = row.find(".issue-disposition-dropdown").native.css_value("border-left")
+          expect(dropdown_border).to eq("4px solid rgb(205, 32, 38)")
```
Comment (mdbenjam) [inline spec/feature/queue/queue_spec.rb]: These seem like good tests. One question I'd have for Alex is how do you all want to structure queue tests. If there are going to be a lot, it might take a long time to run. It might be more efficient to combine some of these tests. Test 2 & 3 duplicate what test 1 is testing (I think). You could also fail to select a disposition, then succeed which would reduce the test count by one more. Not sure if that's a goal.

File: client/app/queue/SelectDispositionsView.jsx [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+  updateAppealIssue,
+  pushBreadcrumb,
+  highlightMissingFormItems
+} from './QueueActions';
+import { fullWidth } from './constants';
+import DecisionViewFooter from './components/DecisionViewFooter';
+
+const marginBottom = (margin) => css({ marginBottom: `${margin}rem` });
+const marginLeft = (margin) => css({ marginLeft: `${margin}rem` });
+const rowStyling = css({
```
[the remark this answers is not in the document]
Comment (dannysellers) [reply client/app/queue/SelectDispositionsView.jsx]: The issue here is that we're rendering each issue as a separate `IssueList` (cause they're inside separate table rows). This seemed like the easiest way to pass one style object to the `Table`, rather than passing separate styles for rows and cells. We could pass the `width: 40%` to the first and third columns separately, but the first column, `IssueList`, already feels kind of overloaded.

File: client/app/queue/components/DecisionViewBase.jsx [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    render = () => <React.Fragment>
+      <Breadcrumbs />
+      <AppSegment filledBackground>
+        <ComponentToWrap ref={this.getWrappedComponentRef} {...this.passthroughProps()} />
+      </AppSegment>
+      <DecisionViewFooter buttons={this.getFooterButtons()} />
+    </React.Fragment>;
+  }
+
+  WrappedComponent.displayName = `DecisionViewBase(${getDisplayName(WrappedComponent)})`;
```
[the remark this answers is not in the document]
Comment (dannysellers) [reply client/app/queue/components/DecisionViewBase.jsx]: [The docs](https://reactjs.org/docs/react-component.html#displayname) say it's for debugging, but it's not something I've ever noticed. I just copied this from the HOC example in the docs, but it makes sense why we'd want to explicitly name a HOC

File: client/app/queue/reducers.js (hunk 2 above)
Comment (dannysellers) [reply client/app/queue/reducers.js]: This was initially motivated by a thought that it might be easiest to modify (a copy of) an appeal in place, which would get POSTed in its entirety, rather than creating separate `editingIssue` or `editingAppeal` type objects in the store. That does start to break down in cases like this one—I'm open to suggestions.

Once thing I'm unsure of is the scope of what we'll potentially have to modify/track changes on in the FE. At the moment, we only support modifying appeals' issues—the OMO checkout flow operates on an object external to the appeal—though we may eventually have to support updating the appeal itself (i.e. judge/attorney assignment)

File: client/app/queue/components/IssueList.jsx [hunk 6]
```diff
@@ -65,20 +69,32 @@ export default class IssueList extends React.PureComponent {
     }
 
     if (issuesOnly) {
-      return <React.Fragment>
-        {appeal.issues.map((issue, idx) => <tr key={`${issue.id}_${issue.vacols_sequence_id}`}>
+      return <React.Fragment>{appeal.issues.map((issue, idx) =>
+        <tr key={`${issue.id}_${issue.vacols_sequence_id}`} {...bottomBorder(singleIssue)}>
           <td {...leftAlignTd}>
-            {idx + 1}.
+            {idxToDisplay || (idx + 1)}.
```
[the remark this answers is not in the document]
Comment (dannysellers) [reply client/app/queue/components/IssueList.jsx]: It doesn't look like there're any issues in Reader, no. If `idxToDisplay` isn't passed, it just renders the row's index. I think we always want to render the number, but if not, we can just give it another prop 😛

**Comment by dannysellers** [comment]
> Right now we have to specify errors twice. Once in the form conditionally to ensure the red highlights appear if we set the `highlightFormItems` flag to true. We also have to write a validation function.

I'm curious if you have any thoughts about how we could consolidate these. In my mind, the next step would be to generify validation and move it into `decisionViewBase`, checking whether a set of fields passed by the wrapped component are present in that component's specific part of the store. I still think we'd need the logic in each form component to conditionally display `errorMessage` though...

File: client/app/queue/components/IssueList.jsx [hunk 7]
```diff
@@ -55,7 +56,10 @@ export default class IssueList extends React.PureComponent {
   getIssues = () => {
     const {
       appeal,
-      issuesOnly
+      issuesOnly,
+      singleIssue,
```
[the remark this answers is not in the document]
Comment (dannysellers) [reply client/app/queue/components/IssueList.jsx]: Briefly discussed this with @laurjpeterson and @Chingujo earlier, they're discussing with other designers soon https://github.com/department-of-veterans-affairs/appeals-design-research/issues/684

File: client/app/queue/SelectDispositionsView.jsx [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+  highlightInvalidFormItems
+} from './QueueActions';
+import { fullWidth } from './constants';
+
+const marginBottom = (margin) => css({ marginBottom: `${margin}rem` });
+const marginLeft = (margin) => css({ marginLeft: `${margin}rem` });
+const tbodyStyling = css({
+  '& > tr': {
+    borderBottom: 'none',
+    '> td': {
```
Comment (mdbenjam) [inline client/app/queue/SelectDispositionsView.jsx]: Another way to do this would be to use: https://github.com/department-of-veterans-affairs/caseflow/blob/master/client/app/reader/DocumentsTable.jsx#L229. But I'm not actually sure that's any better. Let's leave it like this for now.

**Comment by mdbenjam** [comment]
We should not tackle the double validation problem here for sure. I don't have a concrete idea yet of how to solve it, one possibly messy way is to turn parse the wrapped component to look to see if any of the children have errors. We should think about it more.

File: client/app/queue/SelectDispositionsView.jsx (hunk 8 above)
Comment (dannysellers) [reply client/app/queue/SelectDispositionsView.jsx]: That does seem like a clean solution, though the I was under the impression is that we were moving away towards prioritizing glamor when possible ¯\\\_(ツ)_/¯

File: client/app/certification/DocumentsCheck.jsx [hunk 9]
```diff
@@ -59,13 +59,13 @@ export class DocumentsCheck extends React.Component {
         <p>If the document status is marked
           with an <NotFoundIcon />, try checking:</p>
         <ul>The <strong>document type</strong> in VBMS to make sure it's
-          <a href="/certification/help#mismatched-documents"> labeled correctly.</a></ul>
+        <a href="/certification/help#mismatched-documents"> labeled correctly.</a></ul>
```
Comment (mdbenjam) [inline client/app/certification/DocumentsCheck.jsx]: Why are these changed? shouldn't the a's still be indented?

File: client/app/queue/SelectDispositionsView.jsx (hunk 8 above)
Comment (mdbenjam) [reply client/app/queue/SelectDispositionsView.jsx]: Yeah I wonder if you could use it with glamor, but regardless this seems good.

File: client/app/queue/SelectDispositionsView.jsx [hunk 10]
```diff
@@ -88,11 +91,7 @@ class SelectDispositionsView extends React.PureComponent {
   getKeyForRow = (rowNumber) => rowNumber;
   getColumns = () => [{
     header: 'Issues',
-    valueFunction: (issue, idx) => <IssueList
-      appeal={{ issues: [issue] }}
-      singleIssue
-      singleColumn
-      idxToDisplay={idx + 1} />
+    valueFunction: (issue, idx) => <IssueList appeal={{ issues: [issue] }} idxToDisplay={idx + 1} />
```
Comment (mdbenjam) [inline client/app/queue/SelectDispositionsView.jsx]: This seems like a great place to use `IssueListItem` instead.

File: client/app/queue/components/IssueListItem.jsx [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+        {issueContent}
+      </td>
+    </React.Fragment>;
+  };
+}
+
+IssueListItem.propTypes = {
+  issue: PropTypes.object.isRequired,
+  issuesOnly: PropTypes.bool,
+  idxToDisplay: PropTypes.number,
```
Comment (mdbenjam) [inline client/app/queue/components/IssueListItem.jsx]: Why do we need both idx and idxToDisplay?

File: client/app/reader/CommentsTable.jsx [hunk 12]
```diff
@@ -11,11 +11,14 @@ export const getRowObjects = (documents, annotationsPerDocument, searchQuery = '
   const groupedAnnotations = _(annotationsPerDocument).
     map((notes) =>
       notes.map((note) => {
-        const { type, serialized_receipt_date } = documents.filter((doc) => doc.id === note.documentId)[0];
+        const {
+          type: docType,
+          serialized_receipt_date: serializedReceiptDate
```
Comment (mdbenjam) [inline client/app/reader/CommentsTable.jsx]: where does searlizedReceiptDate come from?

Comment (mdbenjam) [reply client/app/reader/CommentsTable.jsx]: Was this change automade my eslint?

Comment (dannysellers) [reply client/app/reader/CommentsTable.jsx]: This was a manual change required by eslint—in `CommentsTable.getRowObjects` (the fn tested here), it was telling me that `serialized_receipt_date` needed to be camel case 😕

File: client/app/intake/reducers/rampElection.js [hunk 13]
```diff
@@ -9,8 +9,8 @@ const formatAppeals = (appeals) => {
     {
       id: appeal.id,
       issues: appeal.issues.map(
-        ({ program_description, ...rest }) => ({
-          programDescription: program_description,
+        ({ program_description: programDescription, ...rest }) => ({
```
Comment (mdbenjam) [inline client/app/intake/reducers/rampElection.js]: I don't see how this works either...

File: client/app/reader/CommentsTable.jsx (hunk 12 above)
Comment (mdbenjam) [reply client/app/reader/CommentsTable.jsx]: But I don't see where serializedReceiptDate is defined?

Comment (dannysellers) [reply client/app/reader/CommentsTable.jsx]: It looks like it's a method in [`Document`](https://github.com/department-of-veterans-affairs/caseflow/blob/master/app/models/document.rb#L189), included in its [`serializable_hash`](https://github.com/department-of-veterans-affairs/caseflow/blob/master/app/models/document.rb#L144) fn

File: client/app/intake/reducers/rampElection.js (hunk 13 above)
Comment (dannysellers) [reply client/app/intake/reducers/rampElection.js]: The issue here was (I believe) that by using `({ program_description })` in the signature, eslint was complaining about declaring a non-camel-case variable. Now, instead of defining `program_description` and returning it within the object as `programDescription`, we just define the one variable.

IMO Github's way of displaying changes stacked vertically makes this a lot harder to understand than seeing it side-by-side:

screen shot 2018-03-12 at 12 07 44 pm

File: client/app/queue/SelectDispositionsView.jsx (hunk 10 above)
Comment (dannysellers) [reply client/app/queue/SelectDispositionsView.jsx]: So, we may not be able to use a separate `IssueListItem` here after all... Part of the reason I refactored `IssueList` to use a table initially was to properly left-align the text next to the row number:

```
<table>
  <tr>

    <IssueListItem>
      <td width="10px">
        1.
      </td>
      <td>
        <span>Program:</span> Compensation
        <span>Issues:</span> ...
      </td>
    </IssueListItem>

  </tr>
</table>
```

This alignment comes from having two `td`s next to each other, and requires the containing element render the contents directly into a `tr`. However, in `SelectDispositionsView`, we use a `Table` and pass it an `IssueListItem`, which gets rendered into a `td` by [`Row`](https://github.com/department-of-veterans-affairs/caseflow/blob/master/client/app/components/Table.jsx#L67), giving us `td`s embedded inside a `td`, which throws a warning (but renders as expected):

`Warning: validateDOMNesting(...): <td> cannot appear as a child of <td>.`:

screen shot 2018-03-12 at 2 12 15 pm

☹️

Comment (dannysellers) [reply client/app/queue/SelectDispositionsView.jsx]: `IssueList` used to be divs or something less rigid, but there was strange formatting behavior at certain screen sizes https://github.com/department-of-veterans-affairs/caseflow/issues/4638
